### PR TITLE
Add new flag for column lineage

### PR DESCRIPTION
**Pull request by y-bruin**
Asset1 
```
/* @bruin
name: example
type: duckdb.sql
materialization:
   type: table

depends:
  - example2

@bruin */

select * from example2;

```

Asset2
```
/* @bruin
name: example2
type: duckdb.sql
materialization:
   type: table

columns:
  - name: id
    type: integer
    description: "Just a number"
    primary_key: true
    checks:
        - name: not_null
        - name: positive
        - name: non_negative
  - name: country
    type: varchar
    description: "the country"
    primary_key: true
    checks:
        - name: not_null
  - name: name
    type: varchar
    update_on_merge: true
    description: "Just a name"
    checks:
        - name: unique
        - name: not_null
#        - name: pattern
#        value: "^[A-Z][a-z]*$"
   @bruin */

select id,name, country from example2;
```

Test
```
bruin internal parse-asset example.sql
```

Output: https://jsonblob.com/1306646592982212608

### CODE REVIEW

File: cmd/internal.go [hunk 1]
```diff
@@ -105,7 +115,7 @@ func (r *ParseCommand) ParsePipeline(assetPath string) error {
 
 	foundPipeline.WipeContentOfAssets()
 
-	js, err := json.Marshal(foundPipeline)
+	js, err := json.MarshalIndent(foundPipeline, "", "  ")
```
Comment (karakanb) [inline cmd/internal.go:118]: I don't think we should do this, these are being passed onto the vscode extension, the smaller the payload the better

Comment (karakanb) [reply cmd/internal.go:118]: you can always pipe this to `jq` to make it look nicer when working

File: cmd/internal.go [hunk 2]
```diff
@@ -147,15 +157,47 @@ func (r *ParseCommand) Run(assetPath string) error {
 
 	foundPipeline.Assets = nil
 
-	js, err := json.Marshal(struct {
+	result := struct {
 		Asset    *pipeline.Asset    `json:"asset"`
+		Lineage  *sqlparser.Lineage `json:"lineage"`
```
Comment (karakanb) [inline cmd/internal.go:162]: this is not what we have in mind, instead we should be using the lineage value to fill out the columns of the asset, `Asset.Columns` field

File: cmd/internal.go [hunk 3]
```diff
@@ -208,3 +229,64 @@ func PatchAsset() *cli.Command {
...
+	if err != nil {
+		return fmt.Errorf("failed to parse column lineage: %w", err)
+	}
+
+	for _, lineageCol := range lineage.Columns {
+		for _, upstream := range lineageCol.Upstream {
+			upstreamAsset := pipe.GetAssetByName(upstream.Table)
+			if upstreamAsset == nil {
+				return fmt.Errorf("upstream asset not found: %s", upstream.Table)
+			}
```
Comment (karakanb) [inline cmd/internal.go:264]: I don't think we should return errors, we should simply skip if we cannot find an upstream. we can add validation to this later on

File: cmd/internal.go [hunk 4]
```diff
@@ -208,3 +229,64 @@ func PatchAsset() *cli.Command {
...
+
+			upstreamCol := upstreamAsset.GetColumnWithName(upstream.Column)
+			if upstreamCol == nil {
+				return fmt.Errorf("column %s not found in asset %s", upstream.Column, upstream.Table)
+			}
```
Comment (karakanb) [inline cmd/internal.go:269]: same goes for this, maybe the upstream did not define the column, we shouldn't error here.

File: cmd/internal.go [hunk 5]
```diff
@@ -208,3 +229,64 @@ func PatchAsset() *cli.Command {
...
+
+			newCol := *upstreamCol
+			newCol.Name = lineageCol.Name
+
+			if col := asset.GetColumnWithName(lineageCol.Name); col == nil {
+				asset.Columns = append(asset.Columns, newCol)
+			} else {
+				*col = newCol
+			}
```
Comment (karakanb) [inline cmd/internal.go:278]: I think we should merge the whole object, not just the name.

the rules are:
- if the asset has defined any field, it wins
- if not, get it from the found upstream column

File: pkg/sqlparser/parser_test.go [hunk 6]
```diff
@@ -213,6 +213,16 @@ func TestSqlParser_Lineage(t *testing.T) {
...
+		{
+			name: "yuvraj test",
+			sql: `
+				SELECT * from example2
+			`,
+			schema: Schema{
+				"example2": {"id": "str", "manager_id": "str"},
+			},
+			want: &Lineage{},
+		},
```
Comment (karakanb) [inline pkg/sqlparser/parser_test.go]: don't forget to delete these

File: cmd/internal.go (hunk 5 above)
Comment (y-bruin) [reply cmd/internal.go]: that means we can remove else condition because column exist. Correct ?

**Comment by albertobruin** [comment]
lgtm

**Pull request merged**